squeal-postgresql-ltree/src/Squeal/PostgreSQL/LTree.hs

```diff
@@ -0,0 +1,412 @@
+
+-- | Does ltree match lquery?
+(~%) :: Operator (null1 PGlquery) (null0 PGltree) ('Null 'PGbool)
+(~%) = unsafeBinaryOp "~"
+
+-- | Does ltree match any lquery in array?
+(%?) :: Operator
+  (null0 PGltree) (null1 ('PGvararray (null2 PGlquery))) ('Null 'PGbool)
+(%?) = unsafeBinaryOp "?"
+
+-- | Does ltree match any lquery in array?
+(?%) :: Operator
+  (null0 ('PGvararray (null1 PGlquery))) (null2 PGltree) ('Null 'PGbool)
+(?%) = unsafeBinaryOp "?"
+
+-- | Does ltree match ltxtquery?
+(%@) :: Operator (null PGltree) (null PGltxtquery) ('Null 'PGbool)
+(%@) = unsafeBinaryOp "@"
+
+-- | Does ltree match ltxtquery?
+(@%) :: Operator (null  PGltxtquery) (null PGltree) ('Null 'PGbool)
+(@%) = unsafeBinaryOp "@"
+
+-- | `(<>)` Concatenates ltree paths.
+instance Semigroup
+  (Expression grp lat with db params from (null PGltree)) where
+    (<>) = unsafeBinaryOp "||"
+instance Monoid
+  (Expression grp lat with db params from (null PGltree)) where
+    mempty = fromString ""
+    mappend = (<>)
+
+-- | Does array contain an ancestor of ltree?
+(@>%) :: Operator
+  (null0 ('PGvararray (null1 PGltree))) (null2 PGltree) ('Null 'PGbool)
+(@>%) = unsafeBinaryOp "@>"
+
+-- | Does array contain an ancestor of ltree?
+(%<@) :: Operator
+  (null0 PGltree) (null1 ('PGvararray (null2 PGltree))) ('Null 'PGbool)
+(%<@) = unsafeBinaryOp "<@"
+
+-- | Does array contain a descendant of ltree?
+(<@%) :: Operator
+  (null0 ('PGvararray (null1 PGltree))) (null2 PGltree) ('Null 'PGbool)
+(<@%) = unsafeBinaryOp "<@"
+
+-- | Does array contain a descendant of ltree?
+(%@>) :: Operator
+  (null0 PGltree) (null1 ('PGvararray (null2 PGltree))) ('Null 'PGbool)
+(%@>) = unsafeBinaryOp "@>"
+
+-- | Does array contain any path matching lquery?
+(&~) :: Operator
+  (null0 ('PGvararray (null1 PGltree))) (null2 PGlquery) ('Null 'PGbool)
+(&~) = unsafeBinaryOp "~"
+
+-- | Does array contain any path matching lquery?
+(~&) :: Operator
+  (null0 PGlquery) (null1 ('PGvararray (null2 PGltree))) ('Null 'PGbool)
+(~&) = unsafeBinaryOp "~"
+
+-- | Does ltree array contain any path matching any lquery?
+(&?) :: Operator
+  (null0 ('PGvararray (null1 PGltree)))
+  (null2 ('PGvararray (null3 PGlquery)))
+  ('Null 'PGbool)
+(&?) = unsafeBinaryOp "?"
+
+-- | Does ltree array contain any path matching any lquery?
+(?&) :: Operator
+  (null0 ('PGvararray (null1 PGlquery)))
+  (null2 ('PGvararray (null3 PGltree)))
+  ('Null 'PGbool)
+(?&) = unsafeBinaryOp "?"
+
+-- | Does array contain any path matching ltxtquery?
+(&@) :: Operator
+  (null0 ('PGvararray (null1 PGltree))) (null2 PGltxtquery) ('Null 'PGbool)
+(&@) = unsafeBinaryOp "@"
+
+-- | Does array contain any path matching ltxtquery?
+(@&) :: Operator
+  (null0 PGltxtquery) (null1 ('PGvararray (null2 PGltree))) ('Null 'PGbool)
+(@&) = unsafeBinaryOp "@"
+
+-- | Returns first array entry that is an ancestor of ltree, or NULL if none.
+(?@>) :: Operator
+  (null0 ('PGvararray ('NotNull PGltree))) (null1 PGltree) ('Null PGltree)
```

Comment:
Why do we know that these ones are `'NotNull`?

Reply:
```
eitan=# select (array [NULL, NULL] :: ltree[]) ?@> (E'a.b.c' :: ltree);
ERROR:  array must not contain nulls
```